bigquery_etl/cli/routine.py

```diff
@@ -0,0 +1,414 @@
+from ..docs import validate_docs
+from ..util.common import project_dirs
+
+ROUTINE_NAME_RE = re.compile(r"^(?P<dataset>[a-zA-z0-9_]+)\.(?P<name>[a-zA-z0-9_]+)$")
+ROUTINE_DATASET_RE = re.compile(r"^(?P<dataset>[a-zA-z0-9_]+)$")
+ROUTINE_FILE_RE = re.compile(
+    r"^.*/([a-zA-Z0-9-]+)/([a-zA-Z0-9_]+)/([a-zA-Z0-9_]+)/"
+    r"(udf\.sql|stored_procedure\.sql)$"
+)
+DEFAULT_UDF_DEPENDENCY_DIR = "udf_js_lib/"
+DEFAULT_GCS_BUCKET = "moz-fx-data-prod-bigquery-etl"
+DEFAULT_GCS_PATH = ""
+DEFAULT_PROJECT_ID = "moz-fx-data-shared-prod"
+
+
+def _routines_matching_name_pattern(pattern, sql_path, project_id):
+    """Return paths to UDFs matching the name pattern."""
+    sql_path = Path(sql_path)
+    if project_id is not None:
+        sql_path = sql_path / project_id
+
+    all_sql_files = Path(sql_path).rglob("*.sql")
+    routine_files = []
+
+    for sql_file in all_sql_files:
+        match = ROUTINE_FILE_RE.match(str(sql_file))
+        if match:
+            project = match.group(1)
+            dataset = match.group(2)
+            routine_name = match.group(3)
+            routine_name = f"{project}.{dataset}.{routine_name}"
+            if fnmatchcase(routine_name, f"*{pattern}"):
+                routine_files.append(sql_file)
+            elif project_id and fnmatchcase(routine_name, f"{project_id}.{pattern}"):
+                routine_files.append(sql_file)
+
+    return routine_files
+
+
+sql_dir_option = click.option(
+    "--sql_dir",
+    help="Path to directory which contains queries.",
+    type=click.Path(file_okay=False),
+    default="sql",
+    callback=is_valid_dir,
+)
+
+project_id_option = click.option(
+    "--project-id",
+    "--project_id",
+    help="GCP project ID",
+    default=None,
+)
+
+
+@click.group(help="Commands for managing routines.")
+@click.pass_context
+def routine(ctx):
+    """Create the CLI group for the routine command."""
+    ctx.ensure_object(dict)
+    ctx.obj["DEFAULT_PROJECT"] = None
+
+
+@click.group(help="Commands for managing mozfun routines.")
+@click.pass_context
+def mozfun(ctx):
+    """Create the CLI group for the mozfun command."""
+    ctx.ensure_object(dict)
+    ctx.obj["DEFAULT_PROJECT"] = "sql/mozfun"
+
+
+@routine.command(
+    help="Create a new routine. Specify whether the routine is a UDF or "
+    "stored procedure by adding a --udf or --stored_prodecure flag: "
+    "bqetl routine create <dataset>.<name> --udf"
+)
+@click.argument("name")
+@sql_dir_option
+@click.option(
+    "--project-id",
+    "--project_id",
+    help="GCP project routine should be part of.",
+    default="moz-fx-data-shared-prod",
+    callback=is_valid_project,
+)
+@click.option("--udf", "-u", is_flag=True, help="Create a new UDF", default=False)
+@click.option(
+    "--stored_procedure",
+    "--stored-procedure",
+    "-p",
+    is_flag=True,
+    help="Create a new stored procedure",
+    default=False,
+)
+@click.pass_context
+def create(ctx, name, sql_dir, project_id, udf, stored_procedure):
```

Comment:
Creating new UDFs would look like: `bqetl routine create udf.test_udf --udf` or `bqetl mozfun create json.test_udf --udf`

Creating new stored procedures would look like: `bqetl routine create udf.test_udf --stored_procedure` or `bqetl mozfun create json.test_udf --stored_procedure`

The default project is shared prod, but the `--project_id` options allows to change that